File: src/main/java/ru/anseranser/peshki/engine/event/GameEvent.java
package ru.anseranser.peshki.engine.event;

public sealed interface GameEvent {
    record DiceRolled(int playerNumber, java.util.List<Integer> values) implements GameEvent {}
    record PawnPlaced(int playerNumber, int pawnNumber, int cellIndex) implements GameEvent {}
    record PawnMoved(int playerNumber, int pawnNumber, int fromCell, int toCell, int steps) implements GameEvent {}
    record PawnKilled(int killerPlayer, int killerPawn, int victimPlayer, int victimPawn) implements GameEvent {}
    record EnteredHome(int playerNumber, int pawnNumber) implements GameEvent {}
    record TurnEnded(int playerNumber, boolean extraTurn) implements GameEvent {}
    record GameWon(int playerNumber) implements GameEvent {}
    record MoveRejected(String reason) implements GameEvent {}
}

File: web/src/engine/events/GameEvent.ts
export type GameEvent =
  | { type: 'DiceRolled'; playerNumber: number; values: number[] }
  | { type: 'PawnPlaced'; playerNumber: number; pawnNumber: number; cellIndex: number }
  | { type: 'PawnMoved'; playerNumber: number; pawnNumber: number; fromCell: number; toCell: number; steps: number }
  | { type: 'PawnKilled'; killerPlayer: number; killerPawn: number; victimPlayer: number; victimPawn: number }
  | { type: 'EnteredHome'; playerNumber: number; pawnNumber: number }
  | { type: 'TurnEnded'; playerNumber: number; extraTurn: boolean }
  | { type: 'GameWon'; playerNumber: number }
  | { type: 'MoveRejected'; reason: string };
